rust_src/src/vterm.rs

```diff
@@ -0,0 +1,584 @@
+//! libvterm facilities
+
+use remacs_macros::lisp_fn;
+
+use libc::{c_char, c_uchar, c_void, size_t};
+
+use std::{cmp, mem};
+
+use crate::{
+    cmds::{forward_char, forward_line},
+    data::Fset,
+    editfns::{goto_pos, line_beginning_position, line_end_position},
+    lisp::ExternalPtr,
+    lisp::{defsubr, LispObject},
+    multibyte::LispStringRef,
+    obarray::intern,
+    remacs_sys::{
+        allocate_vterm, color_to_rgb_string, get_col_offset, refresh_lines, rgb_string_to_color,
+        row_to_linenr, utf8_to_codepoint, vterm_output_read, vterm_screen_callbacks,
+        vterm_screen_set_callbacks, VtermScrollbackLine,
+    },
+
+    remacs_sys::{
+        buf_charpos_to_bytepos, code_convert_string_norecord, del_range, make_string, pvec_type,
+        send_process, vterminal, EmacsInt, Fput_text_property, Lisp_Type, Qbold, Qcursor_type,
+        Qface, Qitalic, Qnil, Qnormal, Qt, Qutf_8, Qvtermp, STRING_BYTES,
+    },
+
+    // libvterm
+    remacs_sys::{
+        vterm_get_size, vterm_input_write, vterm_keyboard_key, vterm_keyboard_unichar, vterm_new,
+        vterm_obtain_screen, vterm_obtain_state, vterm_output_get_buffer_current,
+        vterm_screen_enable_altscreen, vterm_screen_flush_damage, vterm_screen_reset,
+        vterm_screen_set_damage_merge, vterm_set_size, vterm_set_utf8, vterm_state_get_cursorpos,
+        vterm_state_set_default_colors, vterm_state_set_palette_color, VTermColor, VTermDamageSize,
+        VTermKey, VTermModifier, VTermPos, VTermProp, VTermRect, VTermScreenCell, VTermState,
+        VTermValue,
+    },
+    symbols::LispSymbolRef,
+    threads::ThreadState,
+    vectors::length,
+};
+
+pub const MaximumScrollback: usize = 1000;
+
+pub type LispVterminalRef = ExternalPtr<vterminal>;
+
+impl LispVterminalRef {
+    pub fn get_size(self) -> (i32, i32) {
+        let mut height: i32 = 0;
+        let mut width: i32 = 0;
+        unsafe { vterm_get_size((*self).vt, &mut height, &mut width) };
+        (height, width)
+    }
+
+    pub fn set_size(self, rows: i32, cols: i32) {
+        unsafe {
+            vterm_set_size((*self).vt, rows, cols);
```

Comment:
@leonerd Here. When I resize the window while htop is running it only updates the height. But when I exit htop and run it again it also gets the correct width. Maybe the reason is how emacs handles resizing...

Reply:
The issue is caused by the `COLUMNS` and `LINES` environment variable. Unset these before invoking htop and resizing will work.